packages/logging/lokiLogProvider.go

```diff
@@ -98,11 +92,94 @@ func (llp *LokiLogProvider) getLogs(doguName string, linesCount int) ([]logLine,
 	}
 
-	logrus.Debugf("finished loki query; got %d logLines", len(result))
+	logrus.Debugf("finished loading logs for linecount; got %d logLines", len(result))
+
+	return result, nil
+}
+
+// queryLogs queries logs from loki for the given time-window (startDate and endDate).
+// The allowed time-window is max 30 days (limited by loki).
+// Since loki also has a limit of max 5000 log-lines per request, a query can result in multiple request for the loki backend.
+func (llp *LokiLogProvider) queryLogs(doguName string, startDate *time.Time, endDate *time.Time, filter string) ([]logLine, error) {
+	queryEndDate := llp.clock.Now()
+	if endDate != nil {
+		queryEndDate = *endDate
+	}
+
+	queryStartDate := createQueryStartDateFromEndDate(queryEndDate)
+	if startDate != nil {
+		queryStartDate = *startDate
+	}
+
+	result := make([]logLine, 0)
+	limit := defaultQueryLimit
+	for {
+
+		logLines, err := llp.queryLogsFromLoki(doguName, queryStartDate, queryEndDate, filter, limit)
+		if err != nil {
+			return nil, fmt.Errorf("failed to query logs from loki: %w", err)
+		}
+
+		if len(logLines) <= 0 {
+			// no new lines to read => nothing is left
+			break
+		}
+
+		// prepend logs
+		result = append(logLines, result...)
+
+		if len(logLines) < limit {
+			// the query returned fewer lines than requested => nothing is left
+			break
+		}
+
+		// there are still logs to read -> start with the newest log timestamp from the last response
+		queryEndDate = logLines[0].timestamp
+	}
+
+	// because multiple logs can happen at the exact same timestamp and the query is batched over time,
+	// it is possible that consecutive batches contain the same logLines. These need to be removed.
+	result = deduplicateLogLines(result)
+
+	logrus.Debugf("finished querying logs; got %d logLines", len(result))
 
 	return result, nil
 }
 
+func (llp *LokiLogProvider) queryLogsFromLoki(doguName string, startDate time.Time, endDate time.Time, filter string, limit int) ([]logLine, error) {
+	if limit <= 0 {
+		limit = defaultQueryLimit
+	}
+
+	if limit > maxQueryLimit {
+		return nil, fmt.Errorf("the given limit of %d exceeds the maximum limit of %d", limit, maxQueryLimit)
+	}
+
+	query := fmt.Sprintf("{pod=~\"%s.*\"}", doguName)
+
+	if len(strings.TrimSpace(filter)) != 0 {
+		query = fmt.Sprintf("%s |= \"%s\" ", query, filter)
+	}
+
+	logrus.Debugf("running loki query for '%s' from %s to %s with limit %d", doguName, startDate.Format(time.RFC3339), endDate.Format(time.RFC3339), limit)
+	lokiQueryUrl, err := buildLokiQueryUrl(llp.gatewayUrl, query, startDate, endDate, limit)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build loki-query: %v", err)
+	}
+
+	lokiResp, err := llp.doLokiHttpQuery(lokiQueryUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute loki-query: %v", err)
+	}
+
+	logLines, err := extractLogLinesFromLokiResponse(lokiResp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract logs from loki response: %v", err)
```

Comment:
Should it be %v instead of %w? General question: Is this a wrapped error or a new error?